server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -1021,8 +1021,8 @@ private void finish() {
                     assertNull(onFailure.get());
                     assertThat(getTranslog(indexShard).getGeneration().translogFileGeneration,
                         // if rollback happens we roll translog twice: one when we flush a commit before opening a read-only engine
-                        // and one after replaying translog (upto the global checkpoint); otherwise we roll translog once.
-                        either(equalTo(translogGen + 1)).or(equalTo(translogGen + 2)));
+                        // and two or three times after replaying translog (upto the global checkpoint); otherwise we roll translog once.
```

Comment:
We discussed this on AoN, we're now potentially rolling the translog redundantly one additional time during flush.

Reply:
I looked at this, and think we can remove the redundant rollover with the following:

```
diff --cc server/src/main/java/org/elasticsearch/index/translog/Translog.java
index 0a8e38816fa,0a8e38816fa..60375d5b485
--- a/server/src/main/java/org/elasticsearch/index/translog/Translog.java
+++ b/server/src/main/java/org/elasticsearch/index/translog/Translog.java
@@@ -1615,13 -1615,13 +1615,17 @@@ public class Translog extends AbstractI
      }
  
      /**
--     * Roll the current translog generation into a new generation. This does not commit the
++     * Roll the current translog generation into a new generation if it's not empty. This does not commit the
       * translog.
       *
       * @throws IOException if an I/O exception occurred during any file operations
       */
      public void rollGeneration() throws IOException {
          syncBeforeRollGeneration();
++        // no need to roll over when it's empty
++        if (current.totalOperations() == 0) {
++            return;
++        }
          try (Releasable ignored = writeLock.acquire()) {
              ensureOpen();
              try {
```

The reason we have the rollover in the first place is to make sue we no longer need to hold on to certain ops, or that we do not mix ops from different terms in the same generation. For  an empty translog, neither is necessary.

Reply:
This works fine I think. The only adjustment I had to make is to also check if the primary term is still in sync with the current writer in addition to the operations.